### ISSUE TITLE
Apply apptype inheritence to the stage, deploy, rollback json structures

### ISSUE THREAD

**Issue by everestx** [herqles-io/hq-code-deployment#2]
In working with this framework, I'm finding that inheritance among types would be very helpful.  For instance, we have some 3 or 4 very similar apptypes that deviate slightly in each of the task and/or configuration steps.
1. Ruby API - no assets, no background workers
2. Ruby Basic Webapp - with assets, no background workers
3. Ruby Full Webapp - Full service, with assets and background workers
4. Ruby Background Worker - No web frontend at all, just a daemon

All of these app types are very similar in the stage operation: git clone, bundler, tarball and upload.  However, each have a slight deviation in either the stage or deploy tasks (ex: `deploy_path` or `rake assets:precompile`)

This of course is still all possible when creating a stage/deploy using `-f app.json` options in the CLI, but I'm really apt to prevent apps from specifying any custom tasks that could ultimately be shared.

Adding this inheritance would allow different app-types that inherit from parent app-types.  As an example Ruby could be the parent or super type.  Ruby would then be barebones for stage and deploy: git clone, bundle, tarball, upload.
-  ruby-api inherits from ruby and overrides the deploy_variable for deploy_path.
-  ruby-webui inherits from ruby and adds assets compilation to stage and asset symlink to deploy
-  ruby-worker inherits from ruby and add daemon setup for background worker (sysv-init or systemd depending on platform)
-  ruby-fullservice inherits from ruby but adds worker and webui type tasks to do all of the above.

To get started, we could create a method to manage the read of the apptypes and handle the inheritance and deep merge of the json structure.  This could be recursive to allow multiple layers of inheritance, using an `app_type` or `parent_type` key in each of the json structures.

**Comment by rmb938** [comment]
I agree. We may have to move the app types from json files to python files. Doing that is not to difficult and it allows inheritance using the python model. 

```
class BaseAppType(object)

class RubyType(BaseAppType)

class RubyAPIType(RubyType)
```

ect..

Doing it this way we could add more customization for app types. Like doing similar tasks depending on environment.

**Comment by everestx** [comment]
That's a good idea. I'll work in this direction and see what I can come up with. Would you expect to support both json and/or schematics models?
